### PR TITLE
Support bzlmod

### PR DESCRIPTION
**Pull request by kormide**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature (please, look at the "Scope of the project" section in the README.md file)
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## What is the current behavior?

The requirements file is parsed in a python script that generates `requirements.bzl` and passes information from the file into `whl_library` called by `install_deps` within the generated file. In order to declare the wheel repositories in a bzlmod extension, we need to be able to parse that same information in Starlark.

## What is the new behavior?

Can parse a requirements file in Starlark which will be uses in a pip_parse bzlmod extension. This in in a similar vein to how `rules_js` parses a pnpm lockfile in starklark for the `npm_translate_lock` extension.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

## Other information

This isn't currently used, but it's a nice reviewable chunk for the bzlmod work.

### CODE REVIEW

**Comment by kormide** [comment]
fyi: @alexeagle, @Wyverald

**Comment by alexeagle** [comment]
My understanding is that something was added after you wrote the yaml parser, which permits one module extension to see a workspace that was created by another, and so perhaps that means it's possible to fetch some tool (like yq) and then execute it from a module extension.

Perhaps that means we could use an existing requirements.txt parser like from pip or poetry that's likely to handle all the edge cases correctly. I doubt it, but might be worth a bit of poking.

File: python/pip_install/private/test/requirements_parser_tests.bzl
Comment (alexeagle) [inline python/pip_install/private/test/requirements_parser_tests.bzl:67]: do you only need the name for the module extension implementation, and not the hashes? I thought you'd need all the data required to produce a requirements.bzl file, like the hashes for downloading the packages.

OH I remember - the pip_parse rule doesn't actually know how to download anything, it will still end up calling `pip install` on individual packages which are required by the build...

**Review by alexeagle — COMMENTED** [review]
Maybe @hrfuller can help with reviews as well, he has a good memory and wrote the pip_parse stuff.

**Comment by kormide** [comment]
> My understanding is that something was added after you wrote the yaml parser, which permits one module extension to see a workspace that was created by another, and so perhaps that means it's possible to fetch some tool (like yq) and then execute it from a module extension.
> 
> Perhaps that means we could use an existing requirements.txt parser like from pip or poetry that's likely to handle all the edge cases correctly. I doubt it, but might be worth a bit of poking.

Yeah originally I wanted to call `pip_parse` in the nodule extension but have it output an extra JSON file containing the result of parsing requirements that I could easily load in starlark. But where I got stuck was finding some what to load that file from the pip_parse repo within the extension. I'm not sure that's possible. But maybe I could call `module_ctx.execute` to execute a python script that uses the same parsing library? Can a module extension create and own a file like that?

File: python/pip_install/private/test/requirements_parser_tests.bzl
Comment (kormide) [reply python/pip_install/private/test/requirements_parser_tests.bzl:67]: Yeah this is the same requirement info that gets passed to `whl_library`.

**Comment by Wyverald** [comment]
> Yeah originally I wanted to call pip_parse in the module extension but have it output an extra JSON file containing the result of parsing requirements that I could easily load in starlark.

That's not directly possible (you can't declare a repo in an extension and immediately read from that repo), but you can use the trick mentioned by Alex: have an extension declare a `pip_parse` repo, `use_repo` that repo in your module, and then load from it in another extension.

> But maybe I could call module_ctx.execute to execute a python script that uses the same parsing library? Can a module extension create and own a file like that?

Alternatively, this works too. A module extension can create files, but these files are not addressable after the extension impl function finishes.

**Comment by kormide** [comment]
> > Yeah originally I wanted to call pip_parse in the module extension but have it output an extra JSON file containing the result of parsing requirements that I could easily load in starlark.
> 
> That's not directly possible (you can't declare a repo in an extension and immediately read from that repo), but you can use the trick mentioned by Alex: have an extension declare a `pip_parse` repo, `use_repo` that repo in your module, and then load from it in another extension.
> 
> > But maybe I could call module_ctx.execute to execute a python script that uses the same parsing library? Can a module extension create and own a file like that?
> 
> Alternatively, this works too. A module extension can create files, but these files are not addressable after the extension impl function finishes.

Thanks for the clarification. That's interesting that a module context can create temporary files. I think one extension is a better developer experience than two, so I'll look into that approach and see where it leads me.

File: python/pip_install/private/requirements_parser.bzl
Comment (hrfuller) [inline python/pip_install/private/requirements_parser.bzl:74]: I'm not sure it is valid for a constraint portion of a requirement string to start with a semi-colon, but you should add a test case for it anyway. e.g `parse("SomePackage;") == ("SomePackage", "SomePackage;")`

**Review by hrfuller — APPROVED** [review]
Cool use of a state machine, thanks for working on this.

**Comment by groodt** [comment]
Looks good. I've got a different proposal to consider and wondered what you might think.

1. Update rules_python to consume a pip "installation report" instead of a transitively locked requirements.txt https://pip.pypa.io/en/stable/reference/installation-report/
2. Parse the locked json report with https://bazel.build/rules/lib/json#decode

The benefits of the json installation report has a few benefits:
* It has the direct urls
* It captures the environment (sys info) of the locking platform

The direct url info in particular will be helpful in future, because it would enable us to parse the format into `.bzl` http_file or http_archive for retrieval in repository rules using the bazel machinery without having to invoke pip.

**Comment by Wyverald** [comment]
> The direct url info in particular will be helpful in future, because it would enable us to parse the format into .bzl http_file or http_archive for retrieval in repository rules using the bazel machinery without having to invoke pip.

This might be built into Bzlmod in the future, without you guys having to do anything. With the planned lockfiles feature, Bzlmod will store the results of module extension resolution in the lockfile; said result is actually just the names and definitions of the repos that the extension generated.

**Comment by groodt** [comment]
> > The direct url info in particular will be helpful in future, because it would enable us to parse the format into .bzl http_file or http_archive for retrieval in repository rules using the bazel machinery without having to invoke pip.
> 
> This might be built into Bzlmod in the future, without you guys having to do anything. With the planned lockfiles feature, Bzlmod will store the results of module extension resolution in the lockfile; said result is actually just the names and definitions of the repos that the extension generated.

That would be something different to what I mention above. That would be for locking the bazel rules dependency graph, not the dependency graph(s) of resolved python packages from external sources. There is no mechanism for bzlmod to do that presently because we aren’t capturing it presently. Hence my line of questioning.

**Comment by alexeagle** [comment]
I don't think it's ever a good idea for Bazel to manage the lockfile when a language-idiomatic package manager manages the dependencies and constraints. Then you get a different result outside of Bazel and the language tooling gets confused.

Anyhow this PR is just to make a bzlmod module extension for the existing repository rules, not to redesign them.

File: python/pip_install/private/requirements_parser.bzl
Comment (kormide) [reply python/pip_install/private/requirements_parser.bzl:74]: I thought I saw an example of this somewhere, but I'll double check/add a test case.

**Comment by kormide** [comment]
I'm going to hold off on this for now and see if I can just execute a python script in the module extension to do the parsing (using the same library that pip_parse uses).

File: python/pip_install/private/requirements_parser.bzl
Comment (kormide) [reply python/pip_install/private/requirements_parser.bzl:74]: There are some examples here: https://pip.pypa.io/en/stable/reference/requirement-specifiers/#requirement-specifiers
E.g.
```
SomeProject ; sys_platform == 'win32'
```
Looks like I already do have a test case.

**Comment by fmeum** [comment]
> Thanks for the clarification. That's interesting that a module context can create temporary files. I think one extension is a better developer experience than two, so I'll look into that approach and see where it leads me.

@kormide Note that even though `rules_python` would have to define two module extensions, end users would not need to interact with the one that provides the tools (such as `yq`) for the other.

**Comment by kormide** [comment]
> > Thanks for the clarification. That's interesting that a module context can create temporary files. I think one extension is a better developer experience than two, so I'll look into that approach and see where it leads me.
> 
> @kormide Note that even though `rules_python` would have to define two module extensions, end users would not need to interact with the one that provides the tools (such as `yq`) for the other.

That's good to know. I had ruled that out because I thought that users would have to declare both extensions.

I'm going to put up a solution with a single module extension first because I'm close to getting that working, and if it feels like there's too much duplication I can go back to the two extension approach.

**Comment by kormide** [comment]
I repurposed this PR to include the full bzlmod solution.

Commit 1: requirements starlark parser (added support for parsing pip options)
Commit 2: bzlmod extension
Commit 3: bzlmod release support (workspace snippet, BCR app config)

I ended up going with the starlark parsing of requirements as an initial solution here. The alternative is to add an additional module extension that parses the requirements file and outputs something like JSON that starlark can read. In the end, (I think) we need to have the `whl_library` declarations inside of the pip_parse extension, hence the need for either a parser or another extension to load the wheel arguments.

File: .bcr/template.metadata.json
Comment (kormide) [inline .bcr/template.metadata.json:14]: This is a template for an entry that will appear the Bazel Central Registry ([example](https://github.com/bazelbuild/bazel-central-registry/blob/main/modules/aspect_rules_js/metadata.json)).

I wasn't sure who the main maintainers of rules_python are. @rickeylev @groodt are you okay with being listed here?

File: python/pip_install/extract_wheels/parse_requirements_to_bzl.py
Comment (kormide) [inline python/pip_install/extract_wheels/parse_requirements_to_bzl.py:163]: This points calls to `xyz_requirement(...)` to the aliases inside the pip_parse repo, so that we don't need to declare a bunch of `use_repo`s in MODULE.bazel.

**Comment by kormide** [comment]
Just realized that I still need to support registering the python toolchain via bzlmod. The tests are failing on CI because the bzlmod example doesn't use a consistent python version on all platforms (I think it's using whatever the executor has installed), and it's outputting the python version to the vendored requirements file, causing a diff_test to fail.

**Comment by kormide** [comment]
Added a module extension for registering a python toolchain.

File: examples/bzlmod/MODULE.bazel
Comment (fmeum) [inline examples/bzlmod/MODULE.bazel:16]: Does every root module have to manually register a toolchain or is the host Python used if nothing else is specified?

Comment (kormide) [reply examples/bzlmod/MODULE.bazel:16]: In the case of this example, I had to use it to fix the python version because the generated `requirements_lock.txt` file stamps the python version in a comment, and on CI the different platforms had different versions installed, so the diff_test was failing.

File: .bcr/template.metadata.json
Comment (rickeylev) [reply .bcr/template.metadata.json:14]: SGTM

Comment (f0rmiga) [reply .bcr/template.metadata.json:14]: @kormide you can also add me.

File: .bcr/template.metadata.json
Comment (f0rmiga) [inline .bcr/template.metadata.json:13]:
```suggestion
    },
    {
      "name": "Thulio Ferraz Assis",
      "email": "thulio@aspect.dev",
      "github": "f0rmiga"
    }
```

Comment (kormide) [reply .bcr/template.metadata.json:18]: :+1: added

**Comment by kormide** [comment]
This PR needs another review since adding the full bzlmod solution, but it's not clear to me who I should ping for that.

**Comment by f0rmiga** [comment]
@alexeagle is probably the best person to do the final review.

**Review by alexeagle — APPROVED** [review]
nice! sorry it took me a while to get to this

File: .bcr/config.yml
Comment (kormide) [reply .bcr/config.yml:1]: The publish-to-bcr app needs to know who the releaser is in order to:
* send error emails
* tag their username in the bcr PR
* make them the commit author

When a repository uses release automation and the author of a release is the `github-actions` bot (rules_python does this), then the user who caused the release is obscured. `fixedReleaser` also helps in the case where a user is the releaser author, but their email is hidden from their profile, so the app can't access it to send emails or properly attribute the commit to them.

File: .github/workflows/workspace_snippet.sh
Comment (kormide) [reply .github/workflows/workspace_snippet.sh:38]: I found that registering the toolchain outside of bzlmod didn't work. Note that in bazel-lib, we only handle the toolchain internally and it is registered via bzlmod in that repo's `MODULE.bazel`. The user doesn't have to do it but they also can't declare their own version via bzlmod.

File: .bcr/config.yml
Comment (alexeagle) [reply .bcr/config.yml:1]: So if you "push a tag to release", we can never fix this? What if you had to push an annotated tag to release, so that there's a commit-like object in a git ref?

Comment (kormide) [reply .bcr/config.yml:1]: The problem lies with the GitHub api. It uses the identity of whatever authorized token is used to create the release to set as the author. It will not look at annotated tag information and there are no fields to explicitly set the author. So if the actions GITHUB_TOKEN is used to create the release, there's nothing we can really do unless they expand the api.

Comment (alexeagle) [reply .bcr/config.yml:1]: okay well this isn't the PR to dig into that anyhow :)

**Review by alexeagle — APPROVED** [review]

**Pull request merged**
